src/test/java/net/minecraftforge/fml/test/DependencyParserTest.java

```diff
@@ -0,0 +1,220 @@
+        for (DependencyParser parser : parsers)
+        {
+            DependencyParser.DependencyInfo info = parser.parseDependencies("required:" + mod);
+            assertContainsSameToString(info.requirements, Sets.newHashSet(mod));
+            assertTrue(info.dependants.isEmpty());
+            assertTrue(info.dependencies.isEmpty());
+            assertTrue(info.softRequirements.isEmpty());
+        }
+    }
+
+    @Test
+    public void testParsingBefore()
+    {
+        String mod = "supermod3000@[1.2,)";
+        for (DependencyParser parser : parsers)
+        {
+            DependencyParser.DependencyInfo info = parser.parseDependencies("before:" + mod);
+            assertTrue(info.requirements.isEmpty());
+            assertTrue(info.softRequirements.isEmpty());
+            assertContainsSameToString(info.dependants, Sets.newHashSet(mod));
+            assertTrue(info.dependencies.isEmpty());
+        }
+    }
+
+    @Test
+    public void testParsingAfter()
+    {
+        String mod = "supermod3000@[1.2,)";
+        for (DependencyParser parser : parsers)
+        {
+            DependencyParser.DependencyInfo info = parser.parseDependencies("after:" + mod);
+            assertTrue(info.requirements.isEmpty());
+            assertTrue(info.softRequirements.isEmpty());
+            assertTrue(info.dependants.isEmpty());
+            assertContainsSameToString(info.dependencies, Sets.newHashSet(mod));
+        }
+    }
+
+    @Test
+    public void testParsingSoftDependencies()
+    {
+        String mod = "supermod3000@[1.2,)";
+        for (DependencyParser parser : parsers)
+        {
+            DependencyParser.DependencyInfo info = parser.parseDependencies(mod);
+            assertTrue(info.requirements.isEmpty());
+            assertContainsSameToString(info.softRequirements, Sets.newHashSet(mod));
+            assertTrue(info.dependants.isEmpty());
+            assertTrue(info.dependencies.isEmpty());
+        }
+    }
+
+    @Test
+    public void testParsingCombined()
+    {
+        for (DependencyParser parser : parsers)
+        {
+            {
+                DependencyParser.DependencyInfo info = parser.parseDependencies("required:supermod3000@[1.2,)");
+                assertContainsSameToString(info.requirements, Sets.newHashSet("supermod3000@[1.2,)"));
+                assertTrue(info.dependants.isEmpty());
+                assertTrue(info.dependencies.isEmpty());
+                assertTrue(info.softRequirements.isEmpty());
+            }
+
+            {
+                String dependencyString = "after:supermod2000@[1.3,);required-before:yetanothermod;softDepMod@[1.0,2.0);required:modW";
+                DependencyParser.DependencyInfo info = parser.parseDependencies(dependencyString);
+                assertContainsSameToString(info.requirements, Sets.newHashSet("yetanothermod", "modW"));
+                assertContainsSameToString(info.softRequirements, Sets.newHashSet("softDepMod@[1.0,2.0)"));
```

Comment:
Should use all lowercase, as people may reference this test for examples and think that using uppercase mod ids is okay when it isn't.

Reply:
good catch

Reply:
I decided to throw a parsing error when a modId here does not meet the FML requirements, it's too easy to overlook otherwise.

Reply:
On second thought, I tried it with some mod packs and they broke horribly. I will make it log an error or something and have a `TODO` to enable the exception in 1.13.

Reply:
What condition did the mod packs fail on exactly? Anything that runs fine on current 1.11+ should fulfil all the conditions in your sanity check already, shouldn't it?

Reply:
I found 6 mods that depended on modIds with capital letters in them. Currently those deps will do nothing, but if I throw an exception it crashes the pack. I reported the issue to those mods and it should be safe to enable this check in 1.13 eince everyone will need to update and test their mods anyway.

Reply:
Hm, in that case, I think you should be at least logging that the dependency string is invalid. People might have hard dependencies on other mods that will end up crashing the game because they're not enforced at load time.

Reply:
Makes sense, I'll add logging until 1.13.